<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>index.knit</title>

<script src="site_libs/header-attrs-2.14/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">eQTLGen</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
    Home
  </a>
</li>
<li>
  <a href="cohorts.html">
    <span class="fa fa-dna"></span>
     
    Cohorts
  </a>
</li>
<li>
  <a href="projects.html">
    <span class="fa fa-chart-pie"></span>
     
    Projects
  </a>
</li>
<li>
  <a href="documents.html">
    <span class="fa fa-folder"></span>
     
    Documents and guidelines
  </a>
</li>
<li>
  <a href="collaboration.html">
    <span class="fa fa-handshake"></span>
     
    Joining and collaborating
  </a>
</li>
<li>
  <a href="articles.html">
    <span class="fa fa-book"></span>
     
    Articles and preprints
  </a>
</li>
<li>
  <a href="resources.html">
    <span class="fa fa-database"></span>
     
    Resources
  </a>
</li>
<li>
  <a href="https://github.com/eQTLGen">
    <span class="fa fa-github"></span>
     
    
  </a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore"><p><img src="figs/Logo_test2.png"
style="width:3in" /></p></h1>

</div>


<p>The eQTLGen Consortium has been set up to investigate the genetic
architecture of blood gene expression and to understand the genetic
underpinnings of complex traits.</p>
<div id="aims" class="section level2">
<h2>Aims</h2>
<p>The aims of our consortium are:</p>
<ul>
<li>Facilitate the collaboration and data harmonization between
different blood eQTL cohorts.</li>
<li>To take advantage of the high power of consortium and identify weak
genetic effects on blood gene expression.</li>
<li>Provide a framework where consortium members and external
collaborators can pursue their specific scientific interests and to
propose side-projects.</li>
<li>In the form of summary statistics, provide the scientific community
a valuable resource which can be used for interpretation of genome-wide
association studies (GWAS), for the development of novel methods and for
other scientific projects.</li>
<li>Integrate our results with other QTL consortia, in order to
cross-validate the findings and better entangle the molecular mechanisms
leading to the formation of complex traits.</li>
</ul>
</div>
<div id="active-projects" class="section level2">
<h2>Active projects</h2>
<p>❗Currently we are working on the <a
href="projects.html#eqtlgen-phase-ii-global-eqtl-analysis">eQTLGen phase
II analyses</a>: highly-powered <em>genome-wide</em> eQTL meta-analysis
in blood. This project dramatically expands our previous work where we
tested, in addition to <em>cis</em>-eQTLs, <em>trans</em>-eQTLs for a
limited subset of ~10,000 trait-associated SNPs. We plan to conduct such
genome-wide meta-analysis on as many blood-based cohorts as
possible.</p>
<p>Resulting genome-wide association profiles on gene expression would
enable to address many relevant scientific questions. Most simple
example: with such data, you can interpret your GWAS locus by finding
potentially relevant genes from all over the genome, not only in the
near vicinity of trait-associated locus.</p>
<p>Therefore, if you have blood-based eQTL cohort, please consider <a
href="collaboration.html">joining</a> our effort! Please find the
details and analysis cookbook from <a
href="eQTLGen-p2-cookbook.html">here</a>.</p>
<p>If you know any blood-based eQTL cohort not yet joined our
consortium, please drop us a line with the reference and/or contact.</p>
</div>
<div id="previous-work" class="section level2">
<h2>Previous work</h2>
<p>Previously we have conducted <em>cis</em>-eQTL, targeted
<em>trans</em>-eQTL and eQTS analysis in up to 31,684 blood and PBMC
samples from 37 individual cohorts from eQTLGen Consortium (eQTLGen
phase I). In this project, we tested the subset of ~10,000 trait-related
variants for the associations with the expression of distal genes
(<em>trans</em>).</p>
<p>Publication:</p>
<div class="csl-entry">
Võsa &amp; Claringbould et. al. (2021). Large-scale cis- and trans-eQTL
analyses identify thousands of genetic loci and polygenic scores that
regulate blood gene expression. <i>Nature Genetics</i>, <i>53</i>(9),
1300–1310. <a href="https://doi.org/10.1038/s41588-021-00913-z"
class="uri">https://doi.org/10.1038/s41588-021-00913-z</a>
</div>
<p><br> Full summary statistics from eQTLGen phase I are available on
the dedicated <a href="www.eqtlgen.org">web site</a>.</p>
<p>Results from this project have been extensively used in collaboration
projects before and after publication. According to Google Scholar, the
published paper has been cited 100 times and preprint 401 times
(16.06.2022).</p>
</div>
<div id="data" class="section level2">
<h2>Data</h2>
<p>Links to all available summary statistics from our publications and
preprints are under <a href="resources.html">Resources</a></p>
</div>
<div id="administration" class="section level2">
<h2>Administration</h2>
<p>eQTLGen Consortium is jointly coordinated by:</p>
<p>University of Tartu, Institute of Genomics, Estonian Genome Center
(Urmo Võsa, PI: Prof. Tõnu Esko)</p>
<p>University Medical Center Groningen, University of Groningen (Robert
Warmerdam, PI: Prof. Lude Franke)</p>
<p><img src="figs/genoomika%20instituut_eng_hall.svg"
style="width:3in" /> <img
src="figs/rug-umcg_logobalk_enbuitennl_grijsw_staand_a5.jpeg"
style="width:6in" /></p>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
